node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt

```
@@ -901,20 +901,26 @@ internal class SingleThreadedStateMachineManager(
         }
     }
 
-    private fun activeOrRemovedClientIdFuture(existingStatus: FlowWithClientIdStatus, clientId: String) = when (existingStatus) {
-        is FlowWithClientIdStatus.Active -> existingStatus.flowStateMachineFuture
-        is FlowWithClientIdStatus.Removed -> {
-            val flowId = existingStatus.flowId
-            val resultFuture = if (existingStatus.succeeded) {
-                val flowResult = database.transaction { checkpointStorage.getFlowResult(existingStatus.flowId, throwIfMissing = true) }
-                doneFuture(flowResult)
-            } else {
-                val flowException =
-                    database.transaction { checkpointStorage.getFlowException(existingStatus.flowId, throwIfMissing = true) }
-                openFuture<Any?>().apply { setException(flowException as Throwable) }
-            }
+    private fun activeOrRemovedClientIdFuture(
+        existingStatus: FlowWithClientIdStatus,
+        clientId: String,
+        throwIfMissing: Boolean
+    ): CordaFuture<out FlowStateMachineHandle<out Any?>> {
+        return when (existingStatus) {
+            is FlowWithClientIdStatus.Active -> existingStatus.flowStateMachineFuture
+            is FlowWithClientIdStatus.Removed -> {
+                val flowId = existingStatus.flowId
+                val resultFuture = if (existingStatus.succeeded) {
+                    val flowResult = database.transaction { checkpointStorage.getFlowResult(existingStatus.flowId, throwIfMissing) }
+                    doneFuture(flowResult)
+                } else {
+                    val flowException =
+                        database.transaction { checkpointStorage.getFlowException(existingStatus.flowId, throwIfMissing) }
```

What happens when `throwIsMissing` is false and record is not found?  flowException becomes null ?  How does that work in the case below ?

Same applies for `flowResult` above.... can't quite see how null works.

---

This is the scenario we talked about earlier if another thread deleted the result/exception while we are trying to reattach.

In this case they get a `null` future. I guess that is kind of broken then. I need to return `FlowStateMachineHandle<out Any?>?`.

---

Actually I think you're right, we decided if the result doesn't exist that it should throw an exception. This is an edge case anyway.

---

Im not sure whether to throw and exception in this case or not.

`null` is returned if there is no client id matching a flow.
`FlowStateMachineHandle` containing `null` is returned if the result is null.

So I can return `FlowStateMachineHandle` containing `null` if the result/exception is missing but then that overlaps the logic of the result actually being null.

Either I throw an exception, or add a new _optional_ object. `Optional` doesn't suffice for since it cant return `null`s. 

We need somethat says:

- There is a result
- There is a result but it is null
- There is no flow/result matching the client id
- The result was deleted from the db bu there is a matching client id.

So many scenarios....

---

The last one happens due to race conditions, if we think it shouldn't happen often, then go with an exception imo.

If we do that, should we use `Optional`? Or go for `null`. Kotlin doesn't really care, but not sure of the prefered route for Java code anymore.